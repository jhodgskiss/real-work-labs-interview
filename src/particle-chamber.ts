class Particle {
    position: number;
    speed: number;

    constructor(position: number, speed: number) {
        this.position = position;
        this.speed = speed;
    }

    // advance particle and return whether it is still in the chamber
    step(chamberLength: number): boolean {
        this.position += this.speed;
        return this.speed > 0 ? this.position < chamberLength : this.position >= 0;
    }
}

class ParticleChamber {
    particles: Particle[];
    length: number;

    constructor(initialPosition: string, speed: number) {
        this.particles = [];
        this.length = initialPosition.length;

        Array.from(initialPosition).forEach((char, index) => {
            switch (char) {
            case 'R':
                this.addParticle(index, speed);
                break;
            case 'L':
                this.addParticle(index, -speed);
                break;
            default:
                break;
            }
        });
    }

    addParticle(position: number, speed: number) {
        this.particles.push(new Particle(position, speed));
    }

    getChamberState(): string {
        const chamberArray = Array(this.length).fill('.');
        this.particles.forEach(particle => chamberArray[particle.position] = 'X');
        return chamberArray.join('');
    }

    // advances particles and removes those outside chamber
    step() {
        this.particles = this.particles.filter(particle => particle.step(this.length));
    }
}

export function animate(initialPosition: string, speed: number) {
    if (speed === 0) {
        throw new Error('A speed of zero would result in an infinite loop, check inputs.');
    }

    if (!Number.isInteger(speed)) {
        throw new Error('Floats currently not supported.');
    }

    const chamber = new ParticleChamber(initialPosition, speed);
    const chamberArrays: string[] = [chamber.getChamberState()];

    while (chamber.particles.length) {
        chamber.step();
        chamberArrays.push(chamber.getChamberState());
    }

    return chamberArrays;
}
