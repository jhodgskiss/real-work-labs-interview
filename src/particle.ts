class Particle {
    position: number;
    speed: number;

    constructor(position: number, speed: number) {
        this.position = position;
        this.speed = speed;
    }

    getDistanceToExit(chamberLength: number): number {
        return this.speed > 0 ? chamberLength - this.position : this.position;
    }

    step(chamberLength: number): boolean {
        this.position += this.speed;
        return this.speed > 0 ? this.position < chamberLength : this.position >= 0;
    }
}

export function animate(initialPosition: string, speed: number) {
    const particles: Particle[] = [];
    const chamberLength: number = initialPosition.length;

    if (speed === 0) {
        throw new Error('A speed of zero would result in an infinite loop, check inputs.');
    }

    if (!Number.isInteger(speed)) {
        throw new Error('Floats currently not supported.');
    }

    Array.from(initialPosition).forEach((char, index) => {
        switch (char) {
        case 'R':
            particles.push(new Particle(index, speed > 0 ? speed : -speed));
            break;
        case 'L':
            particles.push(new Particle(index, speed > 0 ? -speed : speed));
            break;
        default:
            break;
        }
    });

    const maxDistance = Math.max(...particles.map(particle => particle.getDistanceToExit(chamberLength)), 0);
    const maxSteps = Math.abs(Math.ceil(maxDistance / speed));

    const particleArrays: string[][] = [...Array(maxSteps + 1)].map(() => Array(chamberLength).fill('.'));

    particles.forEach(particle => {
        let time = 0;
        do {
            particleArrays[time++][particle.position] = 'X';
        } while (particle.step(chamberLength));
    });

    return particleArrays.map(charArray => charArray.join(''));
}
