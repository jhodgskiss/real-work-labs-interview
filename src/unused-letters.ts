const _CHARACTER_SET = 'abcdefghijklmnopqrstuvwxyz';
const _CHARACTERS = new Set(Array.from(_CHARACTER_SET));

export function getUnusedCharacters(string: string): string {
    const usedCharacterSet = new Set(Array.from(string.toLowerCase()));
    const unusedCharacters = _CHARACTERS.difference(usedCharacterSet);
    return Array.from(unusedCharacters).join('');
}
