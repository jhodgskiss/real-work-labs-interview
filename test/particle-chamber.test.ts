import { animate } from "../src/particle-chamber";

test('moves left, .....L, 1', () => {
    expect(animate('.....L', 1)).toMatchSnapshot();
});

test('moves left, negative, .....R, -1', () => {
    expect(animate('.....R', -1)).toMatchSnapshot();
});

test('moves right, R....., 2', () => {
    expect(animate('R.....', 2)).toMatchSnapshot();
});

test('moves right, L....., -2', () => {
    expect(animate('L.....', -2)).toMatchSnapshot();
});

test('multiple particles, R.LR.L, 1', () => {
    expect(animate('R.LR.L', 1)).toMatchSnapshot();
});

test('chamber length 1', () => {
    expect(animate('R', 1)).toMatchSnapshot();
});

test('chamber length 1, no particle', () => {
    expect(animate('.', 1)).toMatchSnapshot();
});

test('fast particles, R.LR.L, 20', () => {
    expect(animate('R.LR.L', 20)).toMatchSnapshot();
});

test('negative speeds, R.LR.L, -2', () => {
    expect(animate('R.LR.L', -2)).toMatchSnapshot();
});

test('no chamber length', () => {
    expect(animate('', 1)).toMatchSnapshot();
});

test('errors on 0 speeds', () => {
    expect(() => animate('..R..L.', 0)).toThrowErrorMatchingSnapshot();
});

test('errors on float speeds', () => {
    expect(() => animate('..R..L.', 1.1)).toThrowErrorMatchingSnapshot();
});

test('example 0, ..R...., 2',() => {
    expect(animate('..R....', 2)).toMatchSnapshot();
});

test('example 1, RR..LRL, 3',() => {
    expect(animate('RR..LRL', 3)).toMatchSnapshot();
});

test('example 2, LRLR.LRLR, 2',() => {
    expect(animate('LRLR.LRLR', 2)).toMatchSnapshot();
});

test('example 3, RLRLRLRLRL, 10',() => {
    expect(animate('RLRLRLRLRL', 10)).toMatchSnapshot();
});
