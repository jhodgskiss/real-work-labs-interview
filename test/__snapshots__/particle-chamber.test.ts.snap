// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chamber length 1 1`] = `
[
  "X",
  ".",
]
`;

exports[`chamber length 1, no particle 1`] = `
[
  ".",
]
`;

exports[`errors on 0 speeds 1`] = `"A speed of zero would result in an infinite loop, check inputs."`;

exports[`errors on float speeds 1`] = `"Floats currently not supported."`;

exports[`example 0, ..R...., 2 1`] = `
[
  "..X....",
  "....X..",
  "......X",
  ".......",
]
`;

exports[`example 1, RR..LRL, 3 1`] = `
[
  "XX..XXX",
  ".X.XX..",
  "X.....X",
  ".......",
]
`;

exports[`example 2, LRLR.LRLR, 2 1`] = `
[
  "XXXX.XXXX",
  "X..X.X..X",
  ".X.X.X.X.",
  ".X.....X.",
  ".........",
]
`;

exports[`example 3, RLRLRLRLRL, 10 1`] = `
[
  "XXXXXXXXXX",
  "..........",
]
`;

exports[`fast particles, R.LR.L, 20 1`] = `
[
  "X.XX.X",
  "......",
]
`;

exports[`moves left, .....L, 1 1`] = `
[
  ".....X",
  "....X.",
  "...X..",
  "..X...",
  ".X....",
  "X.....",
  "......",
]
`;

exports[`moves left, negative, .....R, -1 1`] = `
[
  ".....X",
  "....X.",
  "...X..",
  "..X...",
  ".X....",
  "X.....",
  "......",
]
`;

exports[`moves right, L....., -2 1`] = `
[
  "X.....",
  "..X...",
  "....X.",
  "......",
]
`;

exports[`moves right, R....., 2 1`] = `
[
  "X.....",
  "..X...",
  "....X.",
  "......",
]
`;

exports[`multiple particles, R.LR.L, 1 1`] = `
[
  "X.XX.X",
  ".X..X.",
  "X.XX.X",
  "..XX..",
  ".X..X.",
  "X....X",
  "......",
]
`;

exports[`negative speeds, ..R..L.., -2 1`] = `
[
  "..X..X..",
  "X......X",
  "........",
]
`;

exports[`negative speeds, R.LR.L, -2 1`] = `
[
  "X.XX.X",
  ".X..X.",
  "......",
]
`;

exports[`no chamber length 1`] = `
[
  "",
]
`;
